### PR TITLE
Implement cellRtc HLE

### PR DESCRIPTION
**Pull request by clienthax**
Will need a bit more work currently,

Whats left is mainly the parsing and a couple of syscalls, along with tidying up all the copypasta

### CODE REVIEW

File: rpcs3/Emu/Cell/Modules/cellRtc.cpp [hunk 1]
```diff
@@ -3,45 +3,160 @@
...
+	pTick->tick = *nsec / 1000 + *sec * cellRtcGetTickResolution() + UTC_BASE_TICKS;
+
 	return CELL_OK;
 }
 
 s32 cellRtcGetCurrentClock(vm::ptr<CellRtcDateTime> pClock, s32 iTimeZone)
 {
 	cellRtc.todo("cellRtcGetCurrentClock(pClock=*0x%x, time_zone=%d)", pClock, iTimeZone);
 
+	vm::ptr<sys_page_attr_t> page_attr = vm::make_var<sys_page_attr_t>({});
```
Comment (elad335) [inline rpcs3/Emu/Cell/Modules/cellRtc.cpp:76]:
```suggestion
	vm::var<sys_page_attr_t> page_attr({});
```
This is not just a cosmetic change, the previous code was bugged.
And in other places.

**Comment by Nekotekina** [comment]
Hasn't been rebased for ages?

**Comment by clienthax** [comment]
> Hasn't been rebased for ages?

Hmm I fresh cloned this a couple of days ago, not quite sure what happened with github..
Will be an easy rebase atleast as nothing else touches cellrtc
image

File: rpcs3/Emu/Cell/Modules/cellRtc.cpp (hunk 1 above)
Comment (clienthax) [reply rpcs3/Emu/Cell/Modules/cellRtc.cpp]: Done

**Comment by sampletext32** [comment]
Can somebody explain to me, why the "todo" function call is never removed from function very begining? If somebody implements some behaviour, then it's not a "todo", it's "done". If something is done partially, then todo should be made more precise, to exactly point to unimplemented piece. So it can help in future development, to quickly find unimplemented code. And btw, it affects code health.

File: rpcs3/Emu/Cell/Modules/cellRtc.cpp [hunk 2]
```diff
@@ -3,45 +3,211 @@
...
+			STR_CASE(CELL_RTC_ERROR_INVALID_MINUTE);
+			STR_CASE(CELL_RTC_ERROR_INVALID_SECOND);
+			STR_CASE(CELL_RTC_ERROR_INVALID_MICROSECOND);
+		}
+
+		return unknown;
+	});
+}
+
+// Grabbed from JPSCP
```
Comment (JohnHolmesII) [inline rpcs3/Emu/Cell/Modules/cellRtc.cpp:38]:
```suggestion
// Grabbed from JPCSP
```

File: rpcs3/Emu/Cell/Modules/cellRtc.cpp [hunk 3]
```diff
@@ -3,45 +3,211 @@
...
+#include "Emu/Cell/lv2/sys_time.h"
+#include "Emu/Cell/lv2/sys_memory.h"
+#include "Emu/Cell/lv2/sys_ss.h"
+
 LOG_CHANNEL(cellRtc);
 
+template <>
+void fmt_class_string<CellRtcError>::format(std::string& out, u64 arg)
+{
+	format_enum(out, arg, [](auto error) {
```
[the remark this answers is not in the document]
Comment (clienthax) [reply rpcs3/Emu/Cell/Modules/cellRtc.cpp:15]: Every time i change it clang format moves it back ...

**Comment by Nekotekina** [comment]
Use // clang-format off

**Comment by clienthax** [comment]
> Can somebody explain to me, why the "todo" function call is never removed from function very begining? If somebody implements some behaviour, then it's not a "todo", it's "done". If something is done partially, then todo should be made more precise, to exactly point to unimplemented piece. So it can help in future development, to quickly find unimplemented code. And btw, it affects code health.

This is a draft, at least in this case, the TODO's are in place so i can see codepaths easily when debugging.

File: rpcs3/Emu/Cell/Modules/cellRtc.cpp [hunk 4]
```diff
@@ -3,298 +3,1672 @@
 
 #include "cellRtc.h"
 
+#include "Emu/Cell/lv2/sys_time.h"
+#include "Emu/Cell/lv2/sys_memory.h"
+#include "Emu/Cell/lv2/sys_ss.h"
+
+//#include <iomanip>
+//#include <sstream>
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/cellRtc.cpp:11]: why these random commented includes?

File: rpcs3/Emu/Cell/Modules/cellRtc.cpp [hunk 5]
```diff
@@ -3,298 +3,1672 @@
...
+
+// Grabbed from JPCSP
+// This is the # of microseconds between January 1, 0001 and January 1, 1970.
+const u64 RTC_MAGIC_OFFSET = 62135596800000000ULL;
+// This is the # of microseconds between January 1, 0001 and January 1, 1601 (for Win32 FILETIME.)
+const u64 RTC_FILETIME_OFFSET = 50491123200000000ULL;
+
+// 400 years is a convenient number, since leap days and everything cycle every 400 years.
+// 400 years is in other words 20871 full weeks.
+const u64 rtc400YearTicks = 20871ULL * 7ULL * 24ULL * 3600ULL * 1000000ULL;
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/cellRtc.cpp:50]: We usually only use yeetCase for the stuff that is used in cell documentation.
There's a lot of yeetCase going on in this file.

File: rpcs3/Emu/Cell/Modules/cellRtc.cpp [hunk 6]
```diff
@@ -3,298 +3,1672 @@
...
+	ret = cellRtcCheckValid(date_time);
+	if (ret != CELL_OK)
+	{
+		return ret;
+	}
+
+	s32 tzone = iTimeZone;
+
+	s32 weekdayIdx = cellRtcGetDayOfWeek(date_time->year, date_time->month, date_time->day);
+	//	Day name
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/cellRtc.cpp:257]: We don't do tabs, we always use a single space. There's more like this in this file
```suggestion
	// Day name
```

File: rpcs3/Emu/Cell/Modules/cellRtc.cpp [hunk 7]
```diff
@@ -3,298 +3,1672 @@
...
-	cellRtc.todo("cellRtcParseDateTime(pUtc=*0x%x, pszDateTime=%s)", pUtc, pszDateTime);
+	cellRtc.error("cellRtcParseDateTime(pUtc=*0x%x, pszDateTime=%s) -- Implement me", pUtc, pszDateTime);
+
+	std::tm t = {};
+	std::string tz;
+
+
+
+
+	//Below code kinda works
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/cellRtc.cpp:538]: We always use a single space. There's probably more like this in this file
```suggestion
	// Below code kinda works
```

File: rpcs3/Emu/Cell/Modules/cellRtc.cpp [hunk 8]
```diff
@@ -3,298 +3,1672 @@
...
+ Takes a RFC2822 / RFC3339 / asctime String, and converts it to a CellRtcTick
+*/
+error_code cellRtcParseDateTime(vm::ptr<CellRtcTick> pUtc, vm::cptr<char> pszDateTime)
 {
-	cellRtc.todo("cellRtcParseDateTime(pUtc=*0x%x, pszDateTime=%s)", pUtc, pszDateTime);
+	cellRtc.error("cellRtcParseDateTime(pUtc=*0x%x, pszDateTime=%s) -- Implement me", pUtc, pszDateTime);
+
+	std::tm t = {};
+	std::string tz;
+
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/cellRtc.cpp:534]: weird huge empty space

File: rpcs3/Emu/Cell/Modules/cellRtc.cpp [hunk 9]
```diff
@@ -3,298 +3,1672 @@
...
+			{
+				pcVar5 = pcVar5 + 1;
+				if (9 < (pszDateTime[pcVar5] - 0x30U))
+					break;
+				date_time->microsecond += uVar7 * (pszDateTime[pcVar5] + -0x30);
+				if (uVar7 == 100000)
+				{
+					uVar7 = 10000;
+				}
+				else
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/cellRtc.cpp:714]: this huge scoperino can be done with else ifs

File: rpcs3/Emu/Cell/Modules/cellRtc.cpp [hunk 10]
```diff
@@ -3,298 +3,1672 @@
...
+error_code getSecureRtcTime(u64 unk1, u64 unk2, u64 unk3)
+{
+	return sys_ss_secure_rtc(0x3002 /* GET_TIME */, unk1, unk2, unk3); // TODO
+}
+
+// End of internal helper functions
+
+// Helper methods
+
+static bool isLeapYear(u32 year)
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/cellRtc.cpp:95]: We usually only use yeetCase for functions that are used in cell documentation.

File: rpcs3/Emu/Cell/Modules/cellRtc.cpp [hunk 11]
```diff
@@ -3,298 +3,1672 @@
...
+
+	vm::var<CellRtcTick> tick;
+	cellRtcGetTick(pDateTime, tick);
+
+	if (tick->tick < 63082281600000000) //	Max time
+	{
+		ret = CELL_RTC_ERROR_INVALID_VALUE;
+		if (pTick)
+		{
+			ret         = CELL_RTC_ERROR_INVALID_VALUE;
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/cellRtc.cpp:1300]: redundant

File: rpcs3/Emu/Cell/Modules/cellRtc.cpp [hunk 12]
```diff
@@ -3,298 +3,1672 @@
...
+			{
+				pTick->tick = (tick->tick + 0xff1fe2ffc59c6000) / cellRtcGetTickResolution();
+			}
+		}
+		else
+		{
+			ret = CELL_RTC_ERROR_INVALID_VALUE;
+			if (pTick)
+			{
+				ret         = CELL_RTC_ERROR_INVALID_VALUE;
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/cellRtc.cpp:1319]: redundant

File: rpcs3/Emu/Cell/Modules/cellRtc.cpp [hunk 13]
```diff
@@ -3,298 +3,1672 @@
...
+	vm::var<CellRtcTick> tick;
+
+	cellRtcGetTick(pDateTime, tick);
+
+	if (tick->tick < RTC_MAGIC_OFFSET)
+	{
+		ret = CELL_RTC_ERROR_INVALID_VALUE;
+		if (piTime)
+		{
+			ret     = CELL_RTC_ERROR_INVALID_VALUE;
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/cellRtc.cpp:1355]: redundant

File: rpcs3/Emu/Cell/Modules/cellRtc.cpp [hunk 14]
```diff
@@ -3,298 +3,1672 @@
...
+	vm::var<CellRtcTick> tick;
+
+	cellRtcGetTick(pDateTime, tick);
+
+	if (tick->tick < RTC_FILETIME_OFFSET)
+	{
+		ret = CELL_RTC_ERROR_INVALID_VALUE;
+		if (pulWin32FileTime)
+		{
+			ret               = CELL_RTC_ERROR_INVALID_VALUE;
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/cellRtc.cpp:1398]: redundant

File: rpcs3/Emu/Cell/Modules/cellRtc.cpp (hunk 4 above)
Comment (clienthax) [reply rpcs3/Emu/Cell/Modules/cellRtc.cpp:11]: Its for the commented block of parsing code

File: rpcs3/Emu/Cell/Modules/cellRtc.cpp (hunk 9 above)
Comment (clienthax) [reply rpcs3/Emu/Cell/Modules/cellRtc.cpp:674]: Honestly i don't even want to touch this block of code :|

Comment (Megamouse) [reply rpcs3/Emu/Cell/Modules/cellRtc.cpp:674]: All you have to do is remove some scopes and use else if instead of if

Comment (Megamouse) [reply rpcs3/Emu/Cell/Modules/cellRtc.cpp:674]: And to be honest.. all this code does is:
```
if (u_var_7 % 100 == 0)
{
    u_var_7 /= 10;
}
else
{
    u_var_7 = (u_var_7 ^ 10) - 1 >> 0x1f;
}
```

File: rpcs3/Emu/Cell/Modules/cellRtc.cpp [hunk 15]
```diff
@@ -3,298 +3,1551 @@
...
+	{
+		ret = pTick1->tick < pTick0->tick;
+	}
+
+	return not_an_error(ret);
 }
 
-DECLARE(ppu_module_manager::cellRtc)("cellRtc", []()
-{
+DECLARE(ppu_module_manager::cellRtc)
```
Comment (Nekotekina) [inline rpcs3/Emu/Cell/Modules/cellRtc.cpp:1549]:
```suggestion
// clang-format off
// ...
// clang-format on
```

**Pull request merged**